### PR TITLE
Graph

### PR DESCRIPTION
**Pull request by aepaysinger**

### CODE REVIEW

File: graph.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    def __init__(self):
+        self._storage = {}
+
+    def nodes(self):
+        """Returns a list of all the nodes in the graph/"""
+        return [node for node in self._storage]
+
+    def edges(self):
+        """Returns a list of all the edges in the graph."""
+        return [edges for edges in self._storage.values()]
```
Comment (W-Ely) [inline graph.py:14]: These are lists of nodes.

If your graph looks like:
```
{"a": {"b", "c"}}
```
Then your edges would look like:
```
[("a", "b"), ("a", "c"), ("c", "a"), ("b", "a")]
```

File: graph.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
+from linkedlist import LinkedList
+
+
+class Graph:
+    def __init__(self):
+        self._storage = {}
+
+    def nodes(self):
+        """Returns a list of all the nodes in the graph/"""
```
Comment (W-Ely) [inline graph.py]: It's interesting that you have chosen to start writing doc strings...

If you are gong to write them please read this https://peps.python.org/pep-0257/

All of these are written as descriptions but should be commands. Returns --> Return

File: graph.py (hunk 1 above)
Comment (aepaysinger) [reply graph.py:14]: but my graph is like this
{"a": {("a", "b"), ("a", "c")}}

File: graph.py (hunk 2 above)
Comment (aepaysinger) [reply graph.py]: I didn't mean to leave that. thank you!

File: graph.py (hunk 1 above)
Comment (W-Ely) [reply graph.py:14]: I see.  That's probably fine. It might make your traversal logic a little more challenging though.

**Comment by aepaysinger** [comment]
is there anything else i need to change?

File: graph.py (hunk 2 above)
Comment (W-Ely) [reply graph.py]: Bump. Fix these Doc strings or remove them.  Fixing them is preferred so you learn what a doc string should look like, whether you choose to use them or not.

File: graph.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        self._storage[value1].add((value1, value2))
+        self._storage[value2].add((value2, value1))
+
+    def del_node(self, value):
+        """Deletes the node containing 'value' from the graph; raises an error if no such node exists."""
+        if value in self._storage:
+            del self._storage[value]
+            for nodes, edges in self._storage.items():
+                if (nodes, value) in edges:
```
Comment (W-Ely) [inline graph.py]: should nodes be plural here?

File: graph.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            raise ValueError("Value not in graph.")
+
+    def adjacent(self, value1, value2):
+        """Returns True if there is an edge connecting 'value1' and 'value2', False if not; raises an error if either of
+        the supplied values are not in the graph."""
+        try:
+            return (value1, value2) in self._storage[value1] or (
+                value2,
+                value1,
+            ) in self._storage[value2]
```
Comment (W-Ely) [inline graph.py]: Is it possible for it to be in one but not the other?

File: graph.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+
+if __name__ == "__main__":
+    graph = Graph()
+    graph.add_edge(4, 6)
+    graph.add_edge(4, 9)
+    graph.add_edge(20, 6)
+    print(graph.edges())
+    # assert graph._storage == {4: {(4, 6), (4, 9)}, 6: {(6, 4), (6, 20)}, 9: {(9, 4)}, 20: {(20, 6)}}
+    # print(graph.neighbors(6))
```
Comment (W-Ely) [inline graph.py]: .

File: linkedlist.py [hunk 6]
```diff
@@ -70,3 +70,9 @@ class Node:
     def __init__(self, value, next):
         self.value = value
         self.next = next
+
+
+if __name__ == "__main__":
+    # print(LinkedList() == LinkedList())
```
Comment (W-Ely) [inline linkedlist.py:76]: .

File: graph.py [hunk 7]
```diff
@@ -30,52 +31,36 @@ def add_edge(self, value1, value2):
...
-    graph = Graph()
-    graph.add_edge(4, 6)
-    graph.add_edge(4, 9)
-    graph.add_edge(20, 6)
-    print(graph.edges())
-    # assert graph._storage == {4: {(4, 6), (4, 9)}, 6: {(6, 4), (6, 20)}, 9: {(9, 4)}, 20: {(20, 6)}}
-    # print(graph.neighbors(6))
+        """Return if there is an edge connecting 'value1' and 'value2'; raise an error if either of the supplied values are not in graph."""
+        if value1 not in self._storage or value2 not in self._storage:
+            raise ValueError("One of the values are not in the graph")
```
Comment (W-Ely) [inline graph.py:65]: Do you have to check both here to know if an edge is present?

**Review by W-Ely — APPROVED** [review]

**Pull request merged**